### PR TITLE
Trainer heartbeat for async mode

### PR DESCRIPTION
**Pull request by seiriosPlus**
背景：
    全异步模式下，如果有trainer训练出错终止，整个训练不会因为trainer出错而终止。但是有一部分数据不会被训练到导致效果会不及预期。

此PR:
    为训练过程加入心跳模式， 每次trainer跟parameter server发送梯度(参数)时更新一下时间戳， parameter server端会有异步线程定期检测，出现超过指定时长未更新的trainer则停掉整个训练。

收益：
    增强异步训练的效果稳定性。

TODO：
    基于此PR可以做更多的check和提醒，进一步保证异步训练的稳定性。
    1. 部分trainer启动过慢
    2. trainer间参数更新时间差距过大
    3. 部分trainer由于训练数据不均过早结束

### CODE REVIEW

**Comment by seiriosPlus** [comment]
当发现trainer节点有超时未更新梯度(参数)， 0号parameter Server将会报错：

image

File: paddle/fluid/operators/distributed/trainer_heart_beat_monitor.cc [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+#include <ctime>
+
+namespace paddle {
+namespace operators {
+namespace distributed {
+
+DEFINE_int32(trainer_update_interval_secs, 900,
+             " the longest time interval between the trainer update variables");
+
+inline int GetCurrentUS() {
```
Comment (guru4elephant) [inline paddle/fluid/operators/distributed/trainer_heart_beat_monitor.cc]: I think there is a timer in platform::Timer, could you reuse that code or improve upon that code?

File: paddle/fluid/operators/distributed/trainer_heart_beat_monitor.h [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    if (monitor_ == nullptr) {
+      monitor_.reset(new TrainerHeartBeatMonitor(trainers, is_chief, varname));
+    }
+  }
+
+  static std::once_flag init_flag_;
+  static std::unique_ptr<TrainerHeartBeatMonitor> monitor_;
+
+  int trainers_;
+  bool is_chief_;
```
Comment (guru4elephant) [inline paddle/fluid/operators/distributed/trainer_heart_beat_monitor.h]: Did u test the training throughput given this new feature?

File: paddle/fluid/operators/distributed/trainer_heart_beat_monitor.h [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+#include "paddle/fluid/platform/enforce.h"
+
+namespace paddle {
+namespace operators {
+namespace distributed {
+
+enum TrainerStatus { UNINITED = 0, RUNNING, COMPLETED };
+
+struct Trainer {
```
Comment (guru4elephant) [inline paddle/fluid/operators/distributed/trainer_heart_beat_monitor.h]: Please rename this struct as TrainerInfo or other names. Trainer is a very general concept and it is able to do training generally.

File: paddle/fluid/operators/distributed/trainer_heart_beat_monitor.h [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  Trainer() {}
+
+  explicit Trainer(int trainer_id) {
+    this->id = trainer_id;
+    this->status = UNINITED;
+    this->timestamp = 0;
+  }
+};
+
+class TrainerHeartBeatMonitor {
```
Comment (guru4elephant) [inline paddle/fluid/operators/distributed/trainer_heart_beat_monitor.h]: I'm not sure whether this heartBeatMonitor can only be used for Trainer. In general any endpoint can have a HeartBeatMonitor.

File: paddle/fluid/operators/distributed/trainer_heart_beat_monitor.cc (hunk 1 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/distributed/trainer_heart_beat_monitor.cc]: Time is designed for debugging, I can not reuse that code here.

**Review by MrChengmo — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]
lgtm

**Pull request merged**